ml/synthesis/src/scripts/evaluation.py

```diff
@@ -0,0 +1,353 @@
+    async def retrieve_subforest(self, session, levels_left=None):
+        self.children = []
+        if self.is_leaf or levels_left == -1:
+            return
+
+        async with session.get(NITTA_BASEURL + f"/node/{self.sid}/subForest") as resp:
+            children_raw = await resp.json()
+            
+        log_debug(f"{len(children_raw)} children from {self.sid}")
+
+        for child_raw in children_raw:
+            child = NittaNode.from_dict(child_raw)
+            child.parent = self
+            self.children.append(child)
+
+        levels_left_for_child = None if levels_left is None else levels_left - 1
+        await asyncio.gather(
+            *[child.retrieve_subforest(session, levels_left_for_child) for child in self.children]
+        )
+        
+                    
+async def retrieve_whole_nitta_tree(max_depth=None) -> NittaNode:
+    start_time = time.perf_counter()
+    async with ClientSession() as session:
+        async with session.get(NITTA_BASEURL + f"/node/-") as resp:
+            root_raw = await resp.json()
+        root = NittaNode.from_dict(root_raw)
+        await root.retrieve_subforest(session, max_depth)
+    
+    print(f"Finished tree retrieval in {time.perf_counter() - start_time:.2f} s")
+    return root
+
+def _extract_params_dict(node: NittaNode) -> dict:
+    if node.decision.tag in ["BindDecisionView", "DataflowDecisionView"]:
+        result = node.parameters.copy()
+        if node.decision.tag == "DataflowDecisionView":
+            result["pNotTransferableInputs"] = sum(result["pNotTransferableInputs"])
+        return result
+    elif node.decision.tag == "RootView":
+        return {}
+    else:
+        # refactorings
+        return {"pRefactoringType": node.decision.tag}
+
+
+def assemble_tree_dataframe(example: str, node: NittaNode, metrics_distrib=None, include_label=True,
+                            levels_left=None) -> pd.DataFrame:
+    if include_label and metrics_distrib is None:
+        metrics_distrib = node.subtree_leafs_metrics
+
+    self_df = pd.DataFrame(dict(
+        example=example,
+        sid=node.sid,
+        tag=node.decision.tag,
+        old_score=node.score,
+        is_leaf=node.is_leaf,
+        **node.alternative_siblings,
+        **_extract_params_dict(node),
+    ), index=[0])
+    if include_label:
+        self_df["label"] = node.compute_label(metrics_distrib)
+
+    levels_left_for_child = None if levels_left is None else levels_left - 1
+    if node.is_leaf or levels_left == -1:
+        return self_df
+    else:
+        result = [assemble_tree_dataframe(example, child, metrics_distrib, include_label, levels_left_for_child)
+                  for child in node.children]
+        if node.sid != "-":
+            result.insert(0, self_df)
+        return pd.concat(result)
+
+
+dfu = pd.concat([pd.read_csv(d) for d in glob(str(Path(args.data_root) / "*.csv"))]).reset_index(drop=True)
+
+def preprocess_df(df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
already defined in `components`, so we better have it imported it from there and reused :)